### PR TITLE
K Shortest Paths in Python

### PR DESCRIPTION
**Pull request by dqawami**
Added pybind11 implementation of k-shortest-paths to be used in Python. Cleaned up c++ source code in the process

### CODE REVIEW

**Review by uditagar — COMMENTED** [review]
looks fine so far

File: libgraph/src/analytics/k_shortest_paths/ksssp.cpp [hunk 1]
```diff
@@ -87,7 +87,7 @@ bool
 CheckReachabilityAsync(
     GraphTy* graph, const typename GraphTy::Node& source,
     const PushWrap& push_wrap, const EdgeRange& edge_range,
-    uint32_t report_node) {
+    size_t report_node) {
```
Comment (gurbinder533) [inline libgraph/src/analytics/k_shortest_paths/ksssp.cpp]: Shouldn't this be GraphTy::Node as well?

File: libgraph/src/analytics/k_shortest_paths/ksssp.cpp [hunk 2]
```diff
@@ -201,7 +201,7 @@ DeltaStepAlgo(
     const PushWrap& push_wrap, const EdgeRange& edge_range,
     katana::InsertBag<std::pair<Weight, Path*>>* report_paths_bag,
     katana::InsertBag<Path*>* path_pointers, PathAlloc& path_alloc,
-    uint32_t report_node, uint32_t num_paths, uint32_t step_shift) {
+    size_t report_node, size_t num_paths, uint32_t step_shift) {
```
Comment (gurbinder533) [inline libgraph/src/analytics/k_shortest_paths/ksssp.cpp]: ditto

File: libgraph/include/katana/analytics/k_shortest_paths/ksssp.h [hunk 3]
```diff
@@ -6,16 +6,35 @@
 namespace katana::analytics {
 using kSsspPlan = SsspPlan;
 
-enum AlgoReachability { async = 0, syncLevel };
+/// Specifies algorithm used to check path reachability
+class AlgoReachability {
```
Comment (gurbinder533) [inline libgraph/include/katana/analytics/k_shortest_paths/ksssp.h]: Can this be part of the kssspPlan?

File: libgraph/src/analytics/k_shortest_paths/ksssp.cpp [hunk 4]
```diff
@@ -543,45 +538,44 @@ kSSSPWithWrap(
...
-        pg, edge_weight_property_name, start_node, report_node, txn_ctx,
-        algo_reachability, num_paths, step_shift, is_symmetric, plan);
+        pg, edge_weight_property_name, start_node, report_node, num_paths,
+        is_symmetric, txn_ctx, algo_reachability, plan);
   case arrow::Int64Type::type_id:
     return kSSSPWithWrap<int64_t>(
-        pg, edge_weight_property_name, start_node, report_node, txn_ctx,
-        algo_reachability, num_paths, step_shift, is_symmetric, plan);
+        pg, edge_weight_property_name, start_node, report_node, num_paths,
+        is_symmetric, txn_ctx, algo_reachability, plan);
```
Comment (gurbinder533) [inline libgraph/src/analytics/k_shortest_paths/ksssp.cpp]: I think algo_reachbility should be part of the plan.

File: python/katana/local/analytics/_ksssp.pyx [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    :type edge_weight_property_name: str
+    :param edge_weight_property_name: The input property containing edge weights.
+    :type start_node: Node ID
+    :param start_node: The source node
+    :type report_node: Node ID
+    :param report_node: The destination node
+    :type num_paths: int
+    :param num_paths: Number of paths to look for
+    :type is_symmetric: bool
+    :param is_symmetric: Whether or not the path is symmetric. Defaults to false.
```
Comment (gurbinder533) [inline python/katana/local/analytics/_ksssp.pyx]: is_symmetric is for the graph or the path?

**Review by gurbinder533 — COMMENTED** [review]
Overall looks good. Left a few comments.

**Comment by gurbinder533** [comment]
Also, can you please add more CI tests in the CMakeList file in `lonestar/analytics/cpu/k-shortest-paths/CMakeLists.txt`:

- Without edgePropertyType
- With node and edge projection
- For all the algorithms
- For all the reachability algos.

File: python/katana/local/analytics/_ksssp.pyx (hunk 5 above)
Comment (dqawami) [reply python/katana/local/analytics/_ksssp.pyx]: For the graph, thanks for the catch

**Comment by dqawami** [comment]
> Also, can you please add more CI tests in the CMakeList file in `lonestar/analytics/cpu/k-shortest-paths/CMakeLists.txt`:
> 
> * Without edgePropertyType
> * With node and edge projection
> * For all the algorithms
> * For all the reachability algos.

I think edgePropertyName is needed for running analytics functions so the edge property for the graph can be known, any suggestions for how to add a default edge property that will work for any inputted graph?

**Comment by gurbinder533** [comment]
> 

Yes, please look at https://github.com/KatanaGraph/katana/blob/master/libgraph/src/analytics/louvain_clustering/louvain_clustering.cpp#L507. We need similar to this. See if you can pull this out as a utility function.

**Review by gurbinder533 — APPROVED** [review]
Looks good. Please squash all the commit when merging. 
See the example I pasted for adding default weights.

**Review by gurbinder533 — APPROVED** [review]
Looks good.

File: python/katana/local/analytics/_ksssp.pyx [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    :type edge_weight_property_name: str
+    :param edge_weight_property_name: The input property containing edge weights.
+    :type start_node: Node ID
+    :param start_node: The source node
+    :type report_node: Node ID
+    :param report_node: The destination node
+    :type num_paths: int
+    :param num_paths: Number of paths to look for
+    :type is_symmetric: bool
+    :param is_symmetric: Whether or not the graph is symmetric. Defaults to false.
```
Comment (arthurp) [inline python/katana/local/analytics/_ksssp.pyx:182]: This shouldn't be here. I can accept it as a short term fix until we support views, but it's bad.

**Review by arthurp — APPROVED** [review]
Looks reasonable over all.

**Pull request merged**